### PR TITLE
Fix #3222, 3221: A few javalib stream changes

### LINKED ISSUE #3193: javalib Collection does not implement stream() default method
javalib Collection is described by the JVM documentation as having a `stream()` default method.

### LINKED ISSUE #3221: javalib j.u.s.Stream does not implement spliterators method
This is a placeholder to help me remember technical debt and to help others avoid potholes.

Recent work brought `spliterator` support to javalib Arrays & Collections. j.u.stream.Stream instances
inherit the default `spliterator()` method from Collection, but, to follow the Stream API describing
`spliterator` as a terminal operation, need to override with an implementation appropriate to Stream.

That is, somebody is going to try using the `spliterator` method on a j.u.s Stream instance and get
results other than the Stream API requires.

### LINKED ISSUE #3222: javalib j.u.s.Stream & kin need API and implementation work
###### Background

1)  I try to express the same appreciation for the current work as I hope people in the future
     express for my work.  It was suitable to the need and available resources at the time.
     Scala Native has grown since then.

2)  I am working under the assumption that someone, now or in the near future, will attempt to use the SN 
    implementation of javalib j.u.s.Stream & kin.  My guess, for "cheap" auto-parallelization.  Multithread
    support will increase.    

3) Influenced by Scala.js and issues of circular dependencies, there has been a desire to remove
    Scala lib dependencies from Scala Native javalib. Scala language features are OK, but things
    such as scala Collections and Streams are candidates for re-work.  A long story, with lots
    of details.  

   The current SN implementation of j.u.s Stream wraps Scala streams and their Scala 3 replacement LazyLists.
   This makes the implementation a candidate for re-work. 

   However, if it ain't broke, don't fix it.
      

###### API - a.k.a Porcelain 

Trying to use or reading the API show that a number of API methods are not implemented.
Issue #3221 describes one such, there are many more.
  
Another example:  I am not quite sure if there is a way to create an Infinite stream.
The 'obvious' methods seem to be missing. 

###### Implementation

Trying to fix Issue #3221 and/or Issue #3193 reveals limits in the current implementation.
Spliterators have been implemented recently.  They were not available at the
time of the current implementation. Changing the implementation to use them
would allow the implementation to better match the API description. 

A partial list:
1)  The Oracle JVM documentation describes Stream as being based on spliterators.
      We can not view the JVM code, but we can take hints from the docs.

2) Trying to fix Issue #3193 reveals that `StreamSupport` is not implemented.
     `SteamSupport` appears pass its spliterator argument to be used
     lazily by the created stream, rather than destroying laziness by iterating
     over the spliterator at creation.

3) Trying to fix Issue #3221 reveals:

    a) Not all streams are Infinite. Some, such as Stream.of(1, 2, 3) have 
    size information known at the point of creation, Stream.of(1, 2, 3).
    On JVM, that information appears to be retained internally and
    passed to its Stream.spliterator() method.  On SN, that information
    is lost. SN can not be easily made to match the JVM behavior without
    changes in its internals.

    b) The concepts of "Stream pipeline" and its concepts of intermediate
     and terminal operations is pretty fundamental to Streams.
     The current implementation does not handle these concepts.

4) Absolutely no parallelism is implemented. 

    A next evolution of the SN Stream implementation might not be parallel internally
    but should explicitly designed to accommodate it (i.e. internal experimental implementation, 
    promoted and supported in a follow-on evolution).

### PR DESCRIPTION
**Pull request by LeeTibbert**
Fix #3222 Fix #3221 Fix #3193

Provide an evolution of Scala Native support for the Java Stream API.

This evolution is provided to verify correctness and as the basis for further work.
Parallel streams, performance, and huge streams all need addressing.

* All streams  are sequential. 

* Almost all of the (Object) `Steam` class is now implemented.

* Almost all of the `DoubleStream` class is implemented.
     
* `IntStream`  and `LongStream` are _not_ implemented. They
    can be implemented once `Stream` and `DoubleStream` have
    been shown to be correct and robust.

   These class and methods in other classes which depend upon these
   will throw an UnsupportedOperationException.

* Array, Collection, and Optional stream methods are implemented.

* StreamSupport is implemented.

* The `Collector` and `Collectors` classes are implemented.
    + Collectors are quite useful on Scala 3.  Extensive tests 
       demonstrate how to use them idiomatically with lambda expressions.
    + Collectors can be made to work with Scala 2.13 with explicit classes: e.g. `Consumer`, et cetera.
    +  The declarations necessary to use Collectors with Scala 2.12 are likely to make the effort not 
         worth the reward.

* An improved algorithm for `mapMulti` methods is implemented.

 Except where noted in this announcement, all Java 8 through 20 methods are implemented.

##### Notes:

* I have not tried backporting this PR to SN 0.4.  It should prove itself in 0.5.0 before such
   a backport is considered.

* Three files were deleted in this PR: `WrappedScalaStream` and three compatibility variants of `ScalaStream`
   I did not create an entry in `changelog/0.5.0.md` because those files were supposed to be internal to the
   Scala Native build.  The removal should not be visible or a concern to users of javalib.  I can create such
   an entry if reviewers think it prudent.

* In general {`Spliterator`, `Stream`, `Collector`} "Characteristics" are passed around but not used.
   This will need to be remedied in future passes which implement efficient and/or parallel streams().
  - Collector.Characteristics are fully implemented and most fully utilized.

* Creating streams of one AnyRef/Object element on Scala 2.12 requires
  special handling to avoid "Stream.of()" method resolution difficulties.
  'Stream.of("borked")' will fail. The following code from `StreamTest`  is not 
  pretty but works across Scala 2.12, 2.13, and 3.n: 
```
    private def streamOfSingleton[T](single: T): Stream[T] = {
     //     val s = Stream.of(expected)  // Use directly if only on Scala 2.13 and 3.n                                         
                                                                            
    val al = new ArrayList[T](1)
    al.add(single)
    al.stream()
  }
```

* Streams are one of those features which probably have two use groups; those who do not use them
    and consider their existence an annoyance and those who use them daily, if not hourly.

    I offer this stream support for its immediate benefit (i.e. Collectors.toList()) for those who expect
    streams to be present and competent. I also am motivated by the idea that efforts to implement
    parallel() streams will allow us to exercise some `java.util.Concurrent`  methods.

   If one is on Java 12 or above, the `teeing()` Collector is pretty useful. It takes one upstream
   and splits it into two similar downstreams.  For example, one downstream could 
   transform/select/reduce items and the other could count them.

### CODE REVIEW

**Comment by LeeTibbert** [comment]
It is not visible to end users but it should be noted that this implementation removes
the prior dependency on Scala streams.  Thank you, @Ekrich for the motivation.

File: javalib/src/main/scala/java/io/BufferedReader.scala
Comment (WojciechMazur) [inline javalib/src/main/scala/java/io/BufferedReader.scala:135]: In this single case `Option` only protects us from null, however we might match on `null` | `line` here directly to skip unnecessary boxing

File: javalib/src/main/scala/java/util/DoubleSummaryStatistics.scala
Comment (WojciechMazur) [inline javalib/src/main/scala/java/util/DoubleSummaryStatistics.scala:9]: Is there some reason why field names start with `_`? Also they could be made private
```suggestion
  private var count: Long = 0L
  private var min: Double = jl.Double.POSITIVE_INFINITY
  private var max: Double = jl.Double.NEGATIVE_INFINITY
  private var sum: Double = 0.0
```

In fact, there could also be defined within the primary constructor, but it's not crucial

File: javalib/src/main/scala/java/util/stream/Collector.scala
Comment (WojciechMazur) [inline javalib/src/main/scala/java/util/stream/Collector.scala:69]: Recently I've found a way to satisfy both Scala 2 and Scala 3 enums, by using `extends _Enum[T]`. Thanks to the name override (which would be fixed in compiler plugin, similarly to the _Object, _Class), Scala 3 compiler would not give errors. Here's example: https://github.com/scala-native/scala-native/blob/ac411cf5c52ca8febd06d1dad2b2630b911e6fb2/javalib/src/main/scala/java/lang/Thread.scala#L358-L379

This would also make sure that type check `_.isInstanceOf[Enum[_]]` would pass in runtime.

**Review by WojciechMazur — COMMENTED** [review]
Thank you for this great improvement. I'll probably would need more time for proper review, but there are a few things that cought my attention

File: javalib/src/main/scala/java/io/BufferedReader.scala
Comment (LeeTibbert) [reply javalib/src/main/scala/java/io/BufferedReader.scala:135]: @Wojciech 
Thank you for taking a rapid overview look at this PR so soon after it was created.

This is a good suggestion.  Every cycle saved, especially in library code, helps.

I will make the correction, once I get thru CI.

Sorry the eyesores you are bound to encounter on detailed review. 


I have made several refinement passes and hope to continue that once
I have confidence in correctness across hardware and Scala and Java versions.

I always like improving both performance in code and my Scala writing skills,
so call out where iterative improvement is indicated.  Thank you, and I hope
that I do not take up too much of your time.

File: javalib/src/main/scala/java/util/DoubleSummaryStatistics.scala
Comment (LeeTibbert) [reply javalib/src/main/scala/java/util/DoubleSummaryStatistics.scala:9]: The same idiom is used in all three of the SummaryStatistics files, so it is worth my moving it from
"get the silly thing correct" to something presentable.   I believe at the time I was concerned at
a coarser granularity with getting `Collectors` to work.

You mention several things.  I will have to take some time to consider & unpack.

##### re: "private"  
Yes, they should be "private" no matter where they are declared. Good catch.

###### re: underbar in names.  
I think at the time I wrote the first of these I was having 
  problems with conflicts when I had variables and arguments with the same name: understandable.
```
 def this(count: Long, min: Double, max: Double, sum: Double) = {
    this()
    _count = count
    _min = min
    _max = max
    _sum = sum
  }
```
I fell back to other examples in the code base which use "_" to indicate global variables.  
Looking at this, I could try an idiom such as ` this.count = count`. 

If that worked, it would eliminate some astonishment and WT? for reviewers.

###### re: variables in primary constructors.

* I know how one declares  a `var` in a primary  constructor, I do not know how one gives it
   an initial value.  "count" and "sum" will get initialized to the proper 0.0. "min" and "max"
   need values different than the default 0.0.

* Probably 99.9% of the time, the no-args constructor will be used.  My thinking was to reduce
   overhead by having that as the primary constructor.  This imposes some cost on the 4 arg
   constructor in that its `this()` sets variables and then it has to overwrite them. I think the
   amortized cost having the zero-arg constructor will be lower, but I have no concrete data.

Is there something more idiomatic or "normal" to do here?   I am open to suggestions/improvement,
especially since it is  "times three"

File: javalib/src/main/scala/java/util/stream/Collector.scala
Comment (LeeTibbert) [reply javalib/src/main/scala/java/util/stream/Collector.scala:69]: This is the major ugliness.  Thank you for the suggestion.  I will take a run at implementing it.

File: javalib/src/main/scala/java/util/DoubleSummaryStatistics.scala
Comment (LeeTibbert) [reply javalib/src/main/scala/java/util/DoubleSummaryStatistics.scala]: Later, re: _ in variable name.

Whilst working on the Enum recommendation below, I came across the place in `Collector.scala` where
I was having a problem with the Scala 3 (?) compiler not liking argument names and `def` names being 
the same.  Is there a  more idiomatic solution than underbar/underscore?

**Comment by LeeTibbert** [comment]
@WojciechMazur @ekrich 

Do either of you know of a way to have a Test file run in CI when it requires _both_ Scala 3 and, say,
Java 9?

The reason I ask is that I have a series of CollectorTestOnJDKx files which require both Scala3 and
non-Java8.  They run manually with the required versions. I could not figure out how to accomplish
the conjunction of requirements in unit-tests.

The Java N portion of the requirement is a hard requirement. `teeing()` and such only exist on 
certain Java versions.   The Scala 3 requirement comes about because declarations necessary
to use Collectors on Scala 2 make their use  less than  cost-effective.   

I went with the sort term fix of merely holding the tests out of CI and running them privately.
If there is a better fix I would like that.  I am not planning on getting hit by a bus, but I like
to leave hand-off quality work in a public repository.

I could probably make small (days, not months) fixes to unit-tests structure if there was
some consensus on a good way forward.   

I took a quick look during stream development and did not see anything obvious; but then I was 
more concerned with getting the entirety of the stream PR submitted & stable.

**Comment by WojciechMazur** [comment]
@LeeTibbert Currently that would not be possible with current setup, but you can adjust the build settings as follows: 

```diff
diff --git a/project/Settings.scala b/project/Settings.scala
index 32246e296..0b105b901 100644
--- a/project/Settings.scala
+++ b/project/Settings.scala
@@ -422,9 +422,21 @@ object Settings {
     Test / unmanagedSourceDirectories ++= {
       val testsRootDir = baseDirectory.value.getParentFile.getParentFile
       val sharedTestDir = testsRootDir / "shared/src/test"
+      val scalaVersionDir = CrossVersion
+        .partialVersion(scalaVersion.value)
+        .collect {
+          case (3, _)     => "scala3"
+          case (2, minor) => s"scala2.$minor"
+        }
+        .getOrElse(sys.error("Unsupported Scala version"))
+
       // Java 8 is reference so start at 9
-      (9 to (Global / javaVersion).value).map { v =>
-        sharedTestDir / s"require-jdk$v"
+      (9 to (Global / javaVersion).value).flatMap { v =>
+        val jdkVersion = s"jdk$v"
+        Seq(
+          sharedTestDir / s"require-$jdkVersion",
+          sharedTestDir / s"require-$jdkVersion-$scalaVersionDir"
+        )
       }
     }
   )
```

It would then allow to define sources in following directories: 
```bash
sbt:scala-native> show testsJVM3/Test/unmanagedSourceDirectories
[info] * /home/wmazur/projects/scalacenter/scala-native/scala-native/unit-tests/jvm/src/test/scala
[info] * /home/wmazur/projects/scalacenter/scala-native/scala-native/unit-tests/jvm/src/test/scala-3
[info] * /home/wmazur/projects/scalacenter/scala-native/scala-native/unit-tests/jvm/src/test/java
[info] * /home/wmazur/projects/scalacenter/scala-native/scala-native/unit-tests/shared/src/test/require-jdk9
[info] * /home/wmazur/projects/scalacenter/scala-native/scala-native/unit-tests/shared/src/test/require-jdk9-scala3
[info] * /home/wmazur/projects/scalacenter/scala-native/scala-native/unit-tests/shared/src/test/require-jdk10
[info] * /home/wmazur/projects/scalacenter/scala-native/scala-native/unit-tests/shared/src/test/require-jdk10-scala3
[info] * /home/wmazur/projects/scalacenter/scala-native/scala-native/unit-tests/shared/src/test/require-jdk11
[info] * /home/wmazur/projects/scalacenter/scala-native/scala-native/unit-tests/shared/src/test/require-jdk11-scala3
```

I'm not only sure about the convention for naming directories. The one from snippet above is `require-jdk9-scala2.13`, which follows existing jdk convention, but it also could be defined as `scala-3-require-jdk9` to follow sbt convention for cross-compiling sources. Each of them seems reasonable, and personally both of them are fine to me.

**Comment by LeeTibbert** [comment]
Thank you for the rapid and helpful response on joint scala & jdk dependencies.
Next work-cycle, I will try that out and create a PR for your consideration.

I like the `scala-3-require-jdk9` name, because it puts the major dependency
first. Current CollectorsTest for jdk8 is under scala-3, so that would keep them
closer together.  I can go with either name.

I will keep the CollectorsTestOnJDKx files on hold. Or this PR can be
on hold until the scala-jdk PR merges. Gives me an incentive to 
get that later PR done.

**Comment by ekrich** [comment]
Is it possible to just use an `assumeFalse` type setup to only run on the correct Scala and Java version rather than the directory setup?

File: javalib/src/main/scala/java/util/DoubleSummaryStatistics.scala
Comment (WojciechMazur) [reply javalib/src/main/scala/java/util/DoubleSummaryStatistics.scala]: In Scala 3 having the following class is not allowed
```
class Foo(someValue: Int) {
  def someValue(): Int =  someValue
}
```
Actually usage of underscore `_someValue` or other similar form is probably the cleanest solution and the most idiomatic solution.

**Comment by LeeTibbert** [comment]
Eric,  thank you for the suggestion.  I need something that works at compile time. I believe that
`assumeFalse` works at runtime.  Because Scala 2.12 _really_ does not like calling Collectors,
I can not use `LinkTimeInfo`, compilation fails way before SN link time.

**Review by WojciechMazur — COMMENTED** [review]
Looks good to me. If you're no longer working on this PR it can be merged. There are only 2 files that might be adjusted with styling, otherwise I have not found anything what might be currently problematic

**Comment by LeeTibbert** [comment]
Failing lint is a coursier failure.  I re-ran scalafmt on my system and nothing changed, so I think/hope that
restarting the lint task will pass.

I monitor the other jobs for failures

**Review by WojciechMazur — APPROVED** [review]

**Pull request merged**
